File: USACO_Problems/redistributing_gifts/red_gifts.py
import sys

class Cow:
    def __init__(self, cow:int, preferences:list):
        self.cow = cow
        self.preferences = preferences[:preferences.index(self.cow) + 1]


n = int(input())

cows = []
for cow in range(1, n + 1):
    cows.append(Cow(cow, list(map(int, input().split()))))
# print(cows)

G = [[] for i in range(1 + n)]
for cow in cows:
    G[cow.cow] = cow.preferences
# print(G)

isPath = [[False] * (1 + n) for _ in range(1 + n)]
for i in range(1, 1 + n):
    for j in G[i]:
        isPath[i][j] = True

# Transitive clojure.
for k in range(1, 1 + n): # The mid-point is the outermost for loop
    for i in range(1, 1 + n):
        for j in range(1, 1 + n):
            if(isPath[i][k] and isPath[k][j]):
                isPath[i][j] = True

answer = [0] * (1 + n)
for cow in cows:
    c = cow.cow
    for p in cow.preferences:
        if isPath[p][c]: # This cycle exists: c->p->...->c
            # Cow c can hope for the gift p.
            answer[c] = p
            break

for i in range(1, 1 + n):
    print(answer[i])


# import sys

# class Cow:
#     def __init__(self, cow:int, preferences:list):
#         self.cow = cow
#         self.preferences = preferences[:preferences.index(self.cow)]
#         self.preferences.append(self.cow)


# n = int(input())
# cows = []
# G = [[] for i in range(1 + n)]
# switched = [False] * (1 + n)
# reached = [0] * (1 + n)
# answer = []
# for cow in range(1, n + 1):
#     cows.append(Cow(cow, list(map(int, input().split()))))

# print(cows)


# for cow in cows:
#     for pref in cow.preferences:
#         G[cow.cow].append(pref)
    
# print(G)

# def dfs(node):
#     for elm in G[node]:
#         if node == elm:
#             reached[node] = elm
#             switched[node] = True
#         elif not switched[node]:
            

# dfs(1)
# dfs(2)
# dfs(3)
# dfs(4)
# print(reached)

# """
# 1 - 4
# 4 - 6
# 6 - 8
# 8 - 1
# 3 - 3
# 5 - 5
# 2 - 7
# 7 - 2
# """

File: USACO_Problems/redistributing_gifts/red_gifts.cpp
#include <iostream>
#include <vector>
using namespace std;

class Cow {
public:
    int cow;
    vector<int> preferences;

    Cow(int cow, const vector<int>& preferences) {
        this->cow = cow;
        // Copy preferences until cow's number is found
        for(auto& pref : preferences) {
            this->preferences.push_back(pref);
            if(pref == cow) break;
        }
    }
};

int main() {
    int n;
    cin >> n;

    vector<Cow> cows;
    for(int cow = 1; cow <= n; ++cow) {
        vector<int> preferences(n);
        for(int& pref : preferences) {
            cin >> pref;
        }
        cows.push_back(Cow(cow, preferences));
    }

    vector<vector<int>> G(n + 1);
    for(auto& cow : cows) {
        G[cow.cow] = cow.preferences;
    }

    vector<vector<bool>> isPath(n + 1, vector<bool>(n + 1, false));
    for(int i = 1; i <= n; ++i) {
        for(int j : G[i]) {
            isPath[i][j] = true;
        }
    }

    // Compute transitive closure using Floyd-Warshall algorithm
    for(int k = 1; k <= n; ++k) {
        for(int i = 1; i <= n; ++i) {
            for(int j = 1; j <= n; ++j) {
                if(isPath[i][k] && isPath[k][j]) {
                    isPath[i][j] = true;
                }
            }
        }
    }

    vector<int> answer(n + 1, 0);
    for(auto& cow : cows) {
        int c = cow.cow;
        for(int p : cow.preferences) {
            if(isPath[p][c]) { // This cycle exists: c->p->...->c
                answer[c] = p; // Cow c can hope for the gift p.
                break;
            }
        }
    }

    for(int i = 1; i <= n; ++i) {
        cout << answer[i] << endl;
    }

    return 0;
}
